### ISSUE TITLE
Create tests from xdoc examples

### ISSUE THREAD

**Issue by stoyanK7** [checkstyle/checkstyle#13100]
For example,
![Screenshot from 2023-05-21 16-09-17](https://github.com/checkstyle/checkstyle/assets/23459549/0fd4c763-ad6b-4544-b9d5-787dedb5dc91)
should turn into
```java
/*xml
<module name="Checker">
  <module name="TreeWalker">
    <module name="AbstractClassName"/>
  </module>
</module>


*/

package com.puppycrawl.tools.checkstyle.checks.naming.abstractclassname;

// xdoc section -- start
class Test1{
  abstract class AbstractFirstClass {} // OK
  abstract class SecondClass {} // violation
  class AbstractThirdClass {} // violation
  class FourthClass {} // OK
}
// xdoc section -- end
```

---
Edit: As time progressed since this issue was created, a few things changed. Example configurations should be in XML (as shown above) and the code should have delimiters surrounding the class definition. Moreover, examples will be excluded from xdoc-javadoc validation. When this issue is closed, there will be a list of all modules below in `XdocsJavadocsTest`. This list can be removed along with the `if`-statements using it.

For convenience, this issue takes care of multiple things - moving examples to test files, placing macros in place of the examples, adding id to each example, and removing examples from class javadoc.

- [x] AbbreviationAsWordInName
- [x] AbstractClassName (needs anchors)
- [x] AnnotationLocation
- [x] AnnotationOnSameLine
- [x] AnnotationUseStyle
- [x] AnonInnerLength
- [x] ArrayTrailingComma
- [x] ArrayTypeStyle
- [x] AtclauseOrder
- [x] AvoidDoubleBraceInitialization
- [x] AvoidEscapedUnicodeCharacters
- [x] AvoidInlineConditionals
- [x] AvoidNestedBlocks
- [x] AvoidNoArgumentSuperConstructorCall
- [x] AvoidStarImport
- [x] AvoidStaticImport
- [x] BooleanExpressionComplexity
- [x] CatchParameterName
- [x] ClassDataAbstractionCoupling
- [x] ClassFanOutComplexity
- [x] ClassMemberImpliedModifier
- [x] ClassTypeParameterName
- [x] CommentsIndentation
- [x] ConstantName
- [x] CovariantEquals
- [x] CustomImportOrder
- [x] CyclomaticComplexity
- [x] DeclarationOrder
- [x] DefaultComesLast
- [x] DescendantToken
- [x] DesignForExtension
- [x] EmptyBlock
- [x] EmptyCatchBlock
- [x] EmptyForInitializerPad
- [x] EmptyForIteratorPad
- [x] EmptyLineSeparator
- [x] EmptyStatement
- [x] EqualsAvoidNull
- [x] EqualsHashCode
- [x] ExecutableStatementCount
- [x] ExplicitInitialization
- [x] FallThrough
- [x] FileLength
- [x] FileTabCharacter
- [x] FinalClass
- [x] FinalLocalVariable
- [x] FinalParameters
- [x] GenericWhitespace
- [x] Header
- [x] HiddenField
- [x] HideUtilityClassConstructor
- [x] IllegalCatch
- [x] IllegalIdentifierName
- [x] IllegalImport
- [x] IllegalInstantiation
- [x] IllegalThrows
- [x] IllegalToken
- [x] IllegalTokenText
- [x] IllegalType
- [ ] ImportControl
- [x] ImportOrder
- [X] Indentation
- [x] InnerAssignment
- [x] InnerTypeLast
- [x] InterfaceIsType
- [x] InterfaceMemberImpliedModifier
- [x] InterfaceTypeParameterName
- [x] InvalidJavadocPosition
- [x] JavadocBlockTagLocation
- [x] JavadocContentLocation
- [x] JavadocMethod
- [x] JavadocMissingLeadingAsterisk
- [x] JavadocMissingWhitespaceAfterAsterisk
- [x] JavadocPackage
- [x] JavadocParagraph
- [x] JavadocStyle
- [x] JavadocTagContinuationIndentation
- [x] JavadocType
- [x] JavadocVariable
- [x] JavaNCSS
- [x] LambdaBodyLength
- [x] LambdaParameterName
- [x] LeftCurly
- [x] LineLength
- [x] LocalFinalVariableName
- [x] LocalVariableName
- [x] MagicNumber
- [x] MatchXpath
- [x] MemberName
- [x] MethodCount
- [x] MethodLength
- [x] MethodName
- [x] MethodParamPad
- [x] MethodTypeParameterName
- [x] MissingCtor
- [x] MissingDeprecated
- [x] MissingJavadocMethod
- [x] MissingJavadocPackage
- [x] MissingJavadocType
- [x] MissingOverride
- [x] MissingSwitchDefault
- [x] ModifiedControlVariable
- [x] ModifierOrder
- [x] MultipleStringLiterals
- [x] MultipleVariableDeclarations
- [x] MutableException
- [x] NeedBraces
- [x] NestedForDepth
- [x] NestedIfDepth
- [x] NestedTryDepth
- [x] NewlineAtEndOfFile
- [x] NoArrayTrailingComma
- [x] NoClone
- [x] NoCodeInFile
- [x] NoFinalizer
- [x] NoLineWrap
- [x] NonEmptyAtclauseDescription
- [x] NoEnumTrailingComma
- [x] NoWhitespaceAfter
- [x] NoWhitespaceBefore
- [x] NoWhitespaceBeforeCaseDefaultColon
- [x] NPathComplexity
- [x] OneStatementPerLine
- [x] OneTopLevelClass
- [x] OperatorWrap
- [x] OrderedProperties
- [x] OuterTypeFilename
- [x] OuterTypeNumber
- [x] OverloadMethodsDeclarationOrder
- [x] PackageAnnotation
- [x] PackageDeclaration
- [x] PackageName
- [x] ParameterAssignment
- [x] ParameterName
- [x] ParameterNumber
- [x] ParenPad
- [x] PatternVariableName
- [x] RecordComponentNumber
- [x] RecordComponentName
- [x] RecordTypeParameterName
- [x] RedundantImport
- [x] RedundantModifier
- [x] Regexp
- [x] RegexpHeader
- [x] RegexpMultiline
- [x] RegexpOnFilename
- [x] RegexpSingleline
- [x] RegexpSinglelineJava
- [x] RequireThis
- [x] ReturnCount
- [x] RightCurly
- [x] SeparatorWrap
- [x] SimplifyBooleanExpression
- [x] SimplifyBooleanReturn
- [x] SingleLineJavadoc
- [x] RequireEmptyLineBeforeBlockTagGroup
- [x] SingleSpaceSeparator
- [x] StaticVariableName
- [x] StringLiteralEquality
- [x] SummaryJavadoc
- [x] SuperClone
- [x] SuperFinalize
- [x] SuppressWarnings
- [x] ThrowsCount
- [x] TodoComment
- [x] TrailingComment
- [x] Translation
- [x] TypecastParenPad
- [x] TypeName
- [x] UncommentedMain
- [x] UniqueProperties
- [x] UnnecessaryParentheses
- [x] UnnecessarySemicolonInEnumeration
- [x] UnnecessarySemicolonInTryWithResources
- [x] UnnecessarySemicolonAfterOuterTypeDeclaration
- [x] UnnecessarySemicolonAfterTypeMemberDeclaration
- [x] UnusedImports
- [x] UpperEll
- [x] UnusedLocalVariable
- [x] VariableDeclarationUsageDistance
- [x] VisibilityModifier
- [x] WhitespaceAfter
- [x] WhitespaceAround
- [x] WriteTag
- [x] SuppressWarningsHolder

Filters:
- [x] [SeverityMatchFilter](https://checkstyle.org/filters/severitymatchfilter.html)
- [x]     [SuppressionCommentFilter](https://checkstyle.org/filters/suppressioncommentfilter.html)
 - [ ]    [SuppressionFilter](https://checkstyle.org/filters/suppressionfilter.html)
- [x]     [SuppressionSingleFilter](https://checkstyle.org/filters/suppressionsinglefilter.html)
- [ ]     [SuppressionXpathFilter](https://checkstyle.org/filters/suppressionxpathfilter.html)
- [x]     [SuppressionXpathSingleFilter](https://checkstyle.org/filters/suppressionxpathsinglefilter.html)
- [x]     [SuppressWarningsFilter](https://checkstyle.org/filters/suppresswarningsfilter.html)
- [x]     [SuppressWithNearbyCommentFilter](https://checkstyle.org/filters/suppresswithnearbycommentfilter.html)
- [x]     [SuppressWithNearbyTextFilter](https://checkstyle.org/filters/suppresswithnearbytextfilter.html)
- [x]     [SuppressWithPlainTextCommentFilter](https://checkstyle.org/filters/suppresswithplaintextcommentfilter.html)
- [x] [BeforeExecutionExclusionFileFilter](https://checkstyle.org/filefilters/beforeexecutionexclusionfilefilter.html)

**Comment by stoyanK7** [comment]
Where should those examples reside? I suggest we do `~/open-source/checkstyle/src/test/resources/com/puppycrawl/tools/checkstyle/checks/naming/abstractclassname/examples` or `.../xdoc-examples`.

Also, shall we consider using different file names for those? `Example1.java`, `Example2.java` or?

**Comment by nrmancuso** [comment]
My thoughts on filename were to have some numerical component that defines the order that the examples appear. `XDocExample1.java` sounds good to me, in order to make it obvious what each file is for.

**Comment by romani** [comment]
I would put them in separate folder like
src/test/resources/xdoc

Or

src/test/resources-xdoc
Like we did for non compile able.
https://github.com/checkstyle/checkstyle/tree/master/src/it

I still have worries that such inputs might have separate  style rules , like file size less then 20 lines.
some of them will be very exceptional in form.

**Comment by romani** [comment]
We need validation that all Inputs in such folder are executed and used in xdoc.

**Comment by nrmancuso** [comment]
I think it is better to do `src/test/xdoc/resources` and `src/test/xdoc/resources-noncompilable` to easily keep files organized, I agree that we will have additional style/ validation for these as time goes on.

**Comment by romani** [comment]
ok , I agree on this naming 

```
✔ ~/java/github/checkstyle/checkstyle/src/test  
$ tree -L 2
.
├── java
│   └── com
├── xdoc
│   ├── resources
│   └── resources-noncompilable
├── resources
│   ├── archunit.properties
│   ├── com
│   └── simplelogger.properties
└── resources-noncompilable
    └── com
```

or



```
✔ ~/java/github/checkstyle/checkstyle/src/test  
$ tree -L 2
.
├── java
│   └── com
│   ├── resources
│   ├── resources-noncompilable
├── resources
│   ├── archunit.properties
│   ├── simplelogger.properties
│   ├── com
│   └── xdoc
└── resources-noncompilable
    ├── com
    └── xdoc
```

or

```
✔ ~/java/github/checkstyle/checkstyle/src/test  
$ tree -L 2
.
├── java
│   └── com
│   ├── resources
│   ├── resources-noncompilable
├── resources
│   ├── archunit.properties
│   ├── simplelogger.properties
│   └── com
└── resources-noncompilable
│   └── com
├── resources-xdoc
│   └── com
└── resources-xdoc-noncompilable
    └── com

```
https://github.com/checkstyle/checkstyle/tree/master/src/test


-----

if we do not plan to reuse regular test classes, we should make separate folder for xdoc-test as we did for `it`
```
✔ ~/java/github/checkstyle/checkstyle/src
$ tree -L 1
.
├── it
├── main
├── site
├── test
├── xdocs-test
│   ├── java
│   ├── resources
│   └── resources-noncompilable
└── xdocs
```

So they will live by their own life and rules, and might be not a part of code coverage. Better to not keep them as critical part of coverage.
In such layout they match "test" and "it" in structure.

**Comment by nrmancuso** [comment]
I think the last structure proposed makes the most sense, let’s do it.

**Comment by romani** [comment]
@stoyanK7 , we are in agreement, please do first PR to create such structure and all other PRs will not have conflicts and you can fill folders with files concurrently.

@rnveach , please consider proposal on file structure layout.

**Comment by stoyanK7** [comment]
@romani I think this issue shouldn't be closed. We use it as a tracker.

**Comment by stoyanK7** [comment]
@nrmancuso @romani Could you please approve the issue?

**Comment by nrmancuso** [comment]
@stoyanK7 @romani to be clear - all further PRs for this issue will have docs generated from the input files and moved to template form?

**Comment by stoyanK7** [comment]
@nrmancuso I am good with this proposal. Move to tests, make a template with macros, add ids for anchors. It would be easier for me as well.

@romani Please confirm you are okay too.

**Comment by romani** [comment]
> Please confirm you are okay too.

lets do this.

**Comment by nrmancuso** [comment]
@stoyanK7 do we have issue for "Activation of example tests" yet?

**Comment by stoyanK7** [comment]
@nrmancuso Yes, this one https://github.com/checkstyle/checkstyle/issues/13345. I still need to create an example PR to reference in that issue. Shall I prioritize doing that?

**Comment by nrmancuso** [comment]
> Yes, this one https://github.com/checkstyle/checkstyle/issues/13345. 

Long week :) This is what we've been referencing in all the example PR's :)

>Shall I prioritize doing that?

No, let's get existing PRs merged for now

**Comment by nrmancuso** [comment]
@stoyanK7 let's pick some check with examples that use language features > jdk11 to do next, and I will send PR to activate.

**Comment by stoyanK7** [comment]
@nrmancuso `RecordTypeParameterName` or `RecordComponentName`, either is fine, yes

**Comment by nrmancuso** [comment]
@stoyanK7 please update todo list above

**Comment by nrmancuso** [comment]
@stoyanK7 was there something left to do for IllegalIdentifierName?

**Comment by stoyanK7** [comment]
All is done.

There are leftover issues for `ImportControl` at https://github.com/checkstyle/checkstyle/issues/13585 and SuppressionFilter and SuppressionXpathFilter at https://github.com/checkstyle/checkstyle/issues/13606
